ss/pebbledb/db.go

```diff
@@ -354,6 +381,108 @@ func (db *Database) ApplyChangesetAsync(version int64, changesets []*proto.Named
 		Version:    version,
 		Changesets: changesets,
 	}
+
+	if db.config.HashRange > 0 {
+		go func(ver int64) {
+			if err := db.computeMissingRanges(ver); err != nil {
+				fmt.Printf("maybeComputeMissingRanges error: %v\n", err)
+			}
+		}(version)
+	}
+
+	return nil
+}
+
+func (db *Database) computeMissingRanges(latestVersion int64) error {
+	lastHashed, err := db.GetLastRangeHashed()
+	if err != nil {
+		return fmt.Errorf("failed to get last hashed range: %w", err)
+	}
+
+	// Keep filling full chunks until we can't
+	for {
+		nextTarget := lastHashed + db.config.HashRange
+
+		// If we haven't reached the next full chunk boundary yet, stop.
+		// We do NOT do partial chunks.
+		if nextTarget > latestVersion {
+			break
+		}
+
+		// We have a full chunk from (lastHashed+1) .. nextTarget
+		begin := lastHashed + 1
+		end := nextTarget
+
+		if err := db.computeHashForRange(begin, end); err != nil {
+			return err
+		}
+
+		// Mark that we've completed that chunk
+		lastHashed = end
+		if err := db.SetLastRangeHashed(lastHashed); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (db *Database) computeHashForRange(beginBlock, endBlock int64) error {
+	fmt.Printf("Computing hash for chunk [%d, %d]\n", beginBlock, endBlock)
+
+	chunkSize := endBlock - beginBlock + 1
+	if chunkSize <= 0 {
+		// Nothing to do
+		return nil
+	}
+
+	numOfWorkers := int(endBlock/chunkSize) + 1
+
+	for _, moduleName := range util.Modules {
+		dataCh := make(chan types.RawSnapshotNode, 10_000)
+
+		hashCalculator := util.NewXorHashCalculator(chunkSize, numOfWorkers, dataCh)
+
+		go func(mod string) {
+			defer close(dataCh)
+
+			_, err := db.RawIterate(mod, func(key, value []byte, ver int64) bool {
+				// Only feed data whose version is in [beginBlock..endBlock]
+				if ver >= beginBlock && ver <= endBlock {
+					dataCh <- types.RawSnapshotNode{
+						StoreKey: mod,
+						Key:      key,
+						Value:    value,
+						Version:  ver,
+					}
+				}
+				return false
+			})
+			if err != nil {
+				panic(fmt.Errorf("error scanning module %s: %w", mod, err))
```

Comment:
@Kbhat1 if we include dex for historical data, do we need to handle this more gracefully to be able to also process ranges that no longer have the dex store? I don't recall what happens to the underlying module store, whether it will load and have no keys to rawIterate or whether it will just fail?

Reply:
dex shouldn't matter since util.Modules will exclude dex anyway, so we won't compute hash for dex module